### ISSUE TITLE
grml-rescueboot doesn't load 2018.12

### ISSUE THREAD

**Issue by riesebie** [grml/grml#125]
My disk layout:
md0 => vg0 => /boot
md0 => vg0 => /usr
md0 => vg0 => /
md0 => vg0 => .....

I am running rescueboot from Debian sid. Grml isn't able to set up vg0 which is essential for booting the ISO. See also [#918590](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=918590)

**Comment by mika** [comment]
Thanks for forwarding this, @riesebie! I'm wondering though, does an older Grml ISO work for you? How exactly does it fail to boot, any chance you could provide a screenshot/photo of the failing boot sequence?

**Comment by riesebie** [comment]
Thanks for notifying this, @mika! Screenshots can be found [here](https://lxtec.de/grml-pics/). After entering busybox I had to `mdadm --assembe --scan` and `vgchange -ay` to get `vg(0|1)`. `<ctrl-d>` didn't boot anyway. And yes, **2017.05** boots as expected!

**Comment by riesebie** [comment]
Even the latest `grml??-full_sid` from [Daily](http://grml.org/daily/) don't work?

**Comment by mika** [comment]
Thanks for the screenshots and information! Ok, then I assume something within udev/lvm/mdadm seems to have changed and isn't fixed yet, maybe really related to the issue we see in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=918590. :-/

**Comment by riesebie** [comment]
`BOOT_IMAGE=/boot/grml64full/vmlinuz apm=power-off boot=live live-media-path=/live/grml64-full/ bootid=grml64full201812 nomce net.ifnames=0 findiso=/grml/grml64-full_2018.12.iso "  live-media=/dev/sda3,/dev/sdb3,/dev/md0,/dev/mapper/vg0-boot " toram=grml64-full.squashfs`

```` 
ls -al \
total 28
  drwxr-xr-x   15 0        0              420 Jan 24 08:27 .
  drwxr-xr-x   15 0        0              420 Jan 24 08:27 ..
  lrwxrwxrwx    1 0        0                7 Dec 30 14:20 bin -> usr/bin
  -rw-r--r--    1 0        0            18300 Jan 24 08:25 boot.log
  drwxr-xr-x    3 0        0              120 Dec 30 14:20 conf
  drwxr-xr-x    2 0        0               60 Dec 30 14:20 cryptroot
  drwxr-xr-x   16 0        0             3260 Jan 24 08:26 dev
  drwxr-xr-x    8 0        0              340 Dec 30 14:20 etc
  -rwxr-xr-x    1 0        0             6202 Jul 18  2018 init
  lrwxrwxrwx    1 0        0                7 Dec 30 14:20 lib -> usr/lib
  lrwxrwxrwx    1 0        0                9 Dec 30 14:20 lib64 -> usr/lib64
  drwxrwsr-x    6 0        100             90 Jan 24 08:18 mnt
  dr-xr-xr-x  158 0        0                0 Jan 24 08:24 proc
  drwx------    2 0        0               40 Dec 11 11:17 root
  drwxr-xr-x    7 0        0              140 Jan 24 08:26 run
  lrwxrwxrwx    1 0        0                8 Dec 30 14:20 sbin -> usr/sbin
  drwxr-xr-x    9 0        0              260 Dec 30 14:20 scripts
  dr-xr-xr-x   13 0        0                0 Jan 24 08:24 sys
  drwxr-xr-x    2 0        0               40 Jan 24 08:24 tmp
  drwxr-xr-x    6 0        0              120 Dec 30 14:20 usr
  drwxr-xr-x    4 0        0               80 Jan 24 08:24 var 
````

**Comment by mika** [comment]
Thanks @riesebie, this isn't the listing of `/live/grml64-full/` though? :)

**Comment by riesebie** [comment]
Where do I find `/live/grml64-full/`?

**Comment by mika** [comment]
Oh right, that's at the wrong strange of the boot, I just grabbed a coffee, sorry. :)

**Comment by riesebie** [comment]
BTW running grml-rescue on a single disk with /boot on a VG with the very same grml-image runs just fine. So maybe the raid is responsible for the disaster?

**Comment by riesebie** [comment]
Tried to test a [grml-live](https://github.com/grml/grml-live) ISO with latest Debian sid. But this ISO isn't bootable with grml-rescueboot? dd'd to stick it boots fine, though. The ISO was build with 
```# grml-live -s sid -a amd64 -c GRMLBASE,GRML_FULL,AMD64```

**Comment by mika** [comment]
The ISOs generated by grml-live should boot fine, our official ISOs are also built like that, but I assume the issue is related to some initramfs hook that's still bugy in Debian/Grml. Thanks for testing though!

**Comment by mika** [comment]
JFTR, according to @riesebie on grml-devel ML grml64-full 2020.06-rc1 is affected as well (please correct me if I assumed wrong flavor (grml-full vs grml-small) or architecture (64bit vs 32bit)).

**Comment by BruderMussLos** [comment]
> JFTR, according to @riesebie on grml-devel ML grml64-full 2020.06-rc1 is affected as well (please correct me if I assumed wrong flavor (grml-full vs grml-small) or architecture (64bit vs 32bit)).

Can confirm same issue occuring on different RAID Versions with the 2020.06 64bit full build. Noticed this first on a server that has RAID I between two partitions, tested the case on another machine. Although grub works perfectly fine, grml is stuck at the /scripts/init-premount...done line (so the startup of grml itself is affected). Tested it with the Superblock Versions 0.9, 1.0 and 1.2, same symptom everywhere. Added bootoptions swraid, dmraid=on and lvm didn't help, as well as toram (with both of the server racks we wanted to use grml on having more than enough RAM to load the 900 MB iso to).

Did start on an ext2 partition in the test bench tho.

I hope this helps you finding the issue more easily if you haven't already!

**Comment by mika** [comment]
Alright, so this latest comment (which is a f'up on http://ml.grml.org/pipermail/grml/2020-October/011784.html) seems to indicate that it's really only about SW-RAID that's causing this issue. Someone needs to dig deeper into the initramfs hooks/scripts of mdadm.

**Comment by Ionic** [comment]
We've been debugging this a bit and I realized that the mdadm script (`/scripts/local-top/mdadm`) that _should_ be getting executed in the `/lib/live/boot/9990-misc-helpers.sh` script doesn't exist in 2017.05.

This makes sense, since the 2017.05 release was based on Debian stretch/testing, which already used the "newer" mdadm version that replaced the older script with newer version(s) in other directories within the initramfs structure.

As far as I can tell, the affected script needs to be rewritten to use the newer hooks.

What I don't understand is why 2017.05 would work for you. This version should already have been broken, UNLESS you've been using a release candidate which might have had the newer, incompatible mdadm Debian package at that time.

@riesebie can you provide the `/boot/grml*/initrd.img` of the image that works for you so that I can take a look in that initramfs?

**Comment by mika** [comment]
@riesebie so @Ionic digged into this and provided a fix for this issue (many thanks for that!). The daily ISOs as of 2021-03-27 are supposed to include this fix, so feel welcome to give it a try. If you still notice any issues please let us know!

**Comment by Ionic** [comment]
We might need to reopen this.

While it did work in the VM that built the image, there might be an additional bug here that I just hit while testing my custom image on a different machine.

I built my custom image in the VM that I used for testing and which has two virtual disks attached in a RAID 1 setup. This lead to `/etc/mdadm/mdadm.conf` being copied from the host, which included the arrays running on my test VM. Booting this image on a different machine will fail to start up the arrays correctly until `mdadm.conf` is moved away. The problematic part is that this file is baked into the initramfs, so removing it isn't easy for users.

Thinking about how to fix that, currently.
